### PR TITLE
feat: turn into up-to-date CRA template

### LINKED ISSUE #16: Turn into custom CRA template
## Suggested feature

- Turn this package into a [custom template](https://create-react-app.dev/docs/custom-templates) for CRA

## Alternatives

Keep the package as a `yarn create` script

## Potential downsides

- This would be a breaking change

### PR DESCRIPTION
**Pull request by felixjung**
Addresses #issue-number.

## Purpose

This PR updates all SumUp dependencies and turns the project into a custom [CRA template](https://create-react-app.dev/docs/custom-templates). With a CRA template users will be able to bootstrap an app by running `yarn create react-app --template @sumup`. Please refer to the updated README for details.

## Definition of done

- [x] Development completed
- [x] Reviewers assigned

BREAKING CHANGE: This change is similar to a complete rewrite.

### CODE REVIEW

**Review by herberthenrique — APPROVED** [review]
Amazing job @felixjung !

Can we move this project to the sumup-oss ?

**Comment by felixjung** [comment]
Friends, could you please take this for a spin? To do so, check out the branch and run

```
yarn create react-app test-react-app-final --template file:create-sumup-react-app
```

where `create-sumup-react-app` should be the relative path to where you cloned this project.

There is a slight issue with linting. 😓 Because we want to provide a foundry linter config to the bootstrapped project, ESLint picks it up from `template/.eslintrc.js`. There seems to be no flag to disable this behavior, though. Any ideas?

**Comment by felixjung** [comment]
I fixed the ESLint issues and also fixed the test template. Would be good if you gave it another try. Should all work as expected now.

File: .eslintrc.js [hunk 1]
```diff
@@ -6,9 +6,13 @@ module.exports = require('@sumup/foundry/eslint')(
     openSource: true,
   },
   {
+    rules: {
+      'react/jsx-uses-react': 'off',
+      'react/react-in-jsx-scope': 'off',
+    },
```
Comment (connor-baer) [inline .eslintrc.js:12]: Is this thanks to the new JSX pragma?

File: README.md [hunk 2]
```diff
@@ -1,60 +1,77 @@
 <div align="center">
-<h1>⚛️ create-sumup-react-app ⚛️</h1>
+<h1>⚛️ @sumup/cra-template ⚛️</h1>
 
-The easiest way to get started writing a React app with SumUp's standard frontend toolchain.
+A
+[create-react-app template](https://create-react-app.dev/docs/custom-templates)
+for JavaScript React apps with SumUp's frontend stack.
```
Comment (connor-baer) [inline README.md]: Do we want to provide a TypeScript template as well? (see also https://github.com/sumup/create-sumup-react-app/issues/25)

I've thought about _only_ making a TypeScript template available to enforce its adoption. It seems heavy-handed though. What do you think?

File: README.md [hunk 3]
```diff
@@ -1,60 +1,77 @@
...
 2. Navigate to the directory you would like to place your project in.
-3. Run `yarn create @sumup/sumup-react-app my-app`, where `my-app` is the name of your project.
+3. Run `yarn create react-app --template @sumup`, where `my-app` is the name of
+   your project.
 
-This will create the folder `my-app` and initialize a new project inside. The project will be based on [`create-react-app`](https://github.com/facebook/create-react-app) and will use the SumUp's [`Circuit UI`](https://circuit.sumup.com/) component library and [Foundry](https://github.com/sumup/foundry) toolkit.
+This will create the folder `my-app` and initialize a new project inside. The
+project will be based on
+[`create-react-app`](https://github.com/facebook/create-react-app) and will use
+the SumUp's [`Circuit UI`](https://circuit.sumup.com/) component library and
```
Comment (connor-baer) [inline README.md]:
```suggestion
SumUp's [`Circuit UI`](https://circuit.sumup.com/) component library and
```

File: README.md [hunk 4]
```diff
@@ -65,83 +82,107 @@ A linter is a tool that checks your code for:
...
+of instructions here.
 
 #### Visual Studio Code
 
-The ESLint extension for VSCode has a `nodePath` setting. It allows you to add additional paths where the plugin should look when resolving dependencies. By setting it to the default path of Foundry in a project's `node_modules`, you will make VSCode work with Foundry's version of ESLint. Put the following line in your `settings.json`.
+The ESLint extension for VSCode has a `nodePath` setting. It allows you to add
+additional paths where the plugin should look when resolving dependencies. By
+setting it to the default path of Foundry in a project's `node_modules`, you
+will make VSCode work with Foundry's version of ESLint. Put the following line
+in your `settings.json`.
```
Comment (connor-baer) [inline README.md]: I don't have this set in my VS Code and it works just fine. Is this still necessary?

File: README.md [hunk 5]
```diff
@@ -1,60 +1,77 @@
...
   - [Maintainers](#maintainers)
 - [About SumUp](#about-sumup)
 
 ## 💻 Prerequisites
 
-- You should have **[Node.js](https://nodejs.org/)** installed at a version equal or above **`v9.10.0`**. If you cannot do that, see the [Troubleshooting](#troubleshooting) section for ways around this requirement.
-- At the moment this project reqires **[`yarn`](https://yarnpkg.com/)** to be installed on your system. Yarn is a package manager for JavaScript. You can read how to install the Yarn CLI in [their documentation](https://yarnpkg.com/en/docs/install).
-- You will need **`npx`** installed. You can run `yarn global add npx` or `npm install -g npx` to make that happen.
+- You should have **[Node.js](https://nodejs.org/)** installed at a version
+  equal or above **`v14.0.0`**. If you cannot do that, see the
```
Comment (connor-baer) [inline README.md]: What's your motivation for introducing these forced linewraps? 

To be honest, I find that they make the text harder to read and code review becomes more difficult because text shifts across multiple lines when modified 😕

File: prettier.config.js [hunk 6]
```diff
@@ -13,4 +13,20 @@
...
+module.exports = require('@sumup/foundry/prettier')(
+  {},
+  {
+    quoteProps: 'consistent',
+    proseWrap: 'always',
+    overrides: [
+      {
+        files: '*.md',
+        options: { parser: 'markdown' },
+      },
```
Comment (connor-baer) [inline prettier.config.js]: I thought prettier inferred the parser automatically based on the file extension 🤔

File: template/eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+          '**/*.story.*',
+          '**/*.stories.*',
+          '**/setupTests.*',
+          '**/test-utils/*',
+        ],
+        rules: {
+          'import/no-extraneous-dependencies': 'off',
+        },
+      },
+    ],
```
Comment (connor-baer) [inline template/eslintrc.js]: Do you think we should add these to the default config in Foundry? I feel like our default starter projects should be able to use the default config of our toolkit.

File: template/prettier.config.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module.exports = require('@sumup/foundry/prettier')(
+  {},
+  {
+    quoteProps: 'consistent',
```
Comment (connor-baer) [inline template/prettier.config.js]: This is already part of the [default config](https://github.com/sumup-oss/foundry/blob/main/src/configs/prettier/config.ts#L32).

File: template/prettier.config.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    proseWrap: 'always',
```
Comment (connor-baer) [inline template/prettier.config.js]: Again, I'd prefer to change this in Foundry to keep it consistent across projects.

File: template/public/manifest.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
+{
+  "short_name": "React App",
+  "name": "Create React App Sample",
```
Comment (connor-baer) [inline template/public/manifest.json]:
```suggestion
  "short_name": "SumUp App",
  "name": "SumUp React App",
```

File: template/src/index.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+import App from './App';
+import reportWebVitals from './reportWebVitals';
+
+ReactDOM.render(<App />, document.getElementById('root'));
+
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals();
```
Comment (connor-baer) [inline template/src/index.js:12]: I assume this is a CRA feature, right? Would there be a benefit in using our own [signal](https://github.com/sumup-oss/signal) library instead?

File: template/src/setupTests.js [hunk 12]
```diff
@@ -14,6 +15,7 @@ import serializer, { matchers } from 'jest-emotion';
  */
 // eslint-disable-next-line no-undef
 expect.extend(matchers);
+expect.extend(toHaveNoViolations);
```
Comment (connor-baer) [inline template/src/setupTests.js:18]: ❤️

File: template/prettier.config.js (hunk 9 above)
Comment (connor-baer) [reply template/prettier.config.js]: If we use the default Foundry configs, we can also remove the config files from the template and instead run `yarn foundry init` as a `postinstall` script. This would allow the users to pick their own setup.

File: .eslintrc.js (hunk 1 above)
Comment (felixjung) [reply .eslintrc.js:12]: Yes. We no longer need the `React` import at the top and these rules are annoying. Taken from the React team's [blog post](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#eslint) about the new JSX transform.

Guess I could add a comment linking to the post.

File: README.md (hunk 2 above)
Comment (felixjung) [reply README.md]: Yes, my idea would be to work on a TypeScript template and put a big strong warning in the JS template README encouraging the use of TypeScript. The TS template would be installable by `yarn create react-app --template @sumup/typescript`.

File: README.md (hunk 4 above)
Comment (felixjung) [reply README.md]: Not sure. I was mostly not sure which troubleshooting items to keep and which not. I would have liked to just delete all of them. 😅

Regarding the actual problem: I think there may be cases where module resolution fails to pick up foundry's ESLint because multiple versions of ESLint being installed. I had this problem with the pre-4.0 version of Foundry. However, in that case even the CRA build fails, so probably there is no need to do this. I'll get rid of the section.

File: template/prettier.config.js (hunk 9 above)
Comment (felixjung) [reply template/prettier.config.js]: Oh yes, brilliant idea. I only realized that `postinstall` works when trying to fix a problem with the nested ESLint config.

File: README.md (hunk 5 above)
Comment (felixjung) [reply README.md]: Ah, sorry. So, actually I thought I had put this in the prettier config but it seems I did not.

> To be honest, I find that they make the text harder to read

In which context? Isn't it much easier to read prose at <80 character line width? You cannot always count on a view that you see raw markdown in to have soft-wrapping configured. And with prettier configured for Markdown reformatting the code is just a save or pre-commit hook away.

> code review becomes more difficult because text shifts across multiple lines when modified

True, a change causes shifts. At the same time an endless line for an entire paragraph is also somewhat difficult to read.

This is not a hard requirement for me. I'd be very curious to see a demo of the issues you describe, just to be able to understand the problem better.

File: template/prettier.config.js (hunk 8 above)
Comment (felixjung) [reply template/prettier.config.js]: 👏

File: template/src/index.js (hunk 11 above)
Comment (felixjung) [reply template/src/index.js:12]: Yes, this is part of CRA. I think it would be cool to use signal. Would there be any restrictions on how that _can_ be used and how it needs to be configured?

File: template/src/setupTests.js (hunk 12 above)
Comment (felixjung) [reply template/src/setupTests.js:18]: It's in their example. 😅

File: README.md (hunk 5 above)
Comment (connor-baer) [reply README.md]: > In which context? Isn't it much easier to read prose at <80 character line width?

Generally, yes. IMO hard-coding this preference is not the way to go though. 

A common scenario where this breaks is when the editor window is narrower than 80 chars. If softwrap isn't enabled, the content overflows anyway and one needs to scroll again. If softwrap is enabled, then there are orphans (single words on a line) because of the forced line breaks.

Here's what this looks like in my standard editor setup (fullscreen MacBook Pro 15", so not the smallest viewport):

Screenshot of VS Code showing the open README file with orphaned text that is difficult to read

This is simple enough for each user to configure themselves. When the wrapping is done by the editor at "view time", it can intelligently reflow the text for the view context. When the wrapping is hard-coded at "write time", we loose this adaptability.

File: template/prettier.config.js (hunk 9 above)
Comment (connor-baer) [reply template/prettier.config.js]: That's what inspired me 😄

File: template/src/index.js (hunk 11 above)
Comment (connor-baer) [reply template/src/index.js:12]: Hm, yes, the [configuration](https://github.com/sumup-oss/signal#installation) is more involved than I thought. 

TL;DR: You need to generate your own script and inline it.

File: template/prettier.config.js (hunk 9 above)
Comment (felixjung) [reply template/prettier.config.js]: Done

File: README.md (hunk 5 above)
Comment (felixjung) [reply README.md]: I removed the hard wraps and the top-level prettier config override.

File: .eslintrc.js (hunk 1 above)
Comment (felixjung) [reply .eslintrc.js:12]: Opted for leaving the `React` imports and adding a hint to the README.

**Comment by felixjung** [comment]
OK, by opting for running `foundry init` in a `postinstall` hook, I think we lost the ability to easily test this e2e. I've tested locally runing `start`, `build`, `test`, and `create:component` together with `test`. Everything worked. @connor-baer could you take another look?

**Review by connor-baer — APPROVED** [review]
✨
